README.md
![Continuous Integration](https://github.com/Zwirner/design-system/actions/workflows/ci.yml/badge.svg)
# TSDX React w/ Storybook User Guide

Congrats! You just saved yourself hours of work by bootstrapping this project with TSDX. Let’s get you oriented with what’s here and how to use it.

> This TSDX setup is meant for developing React component libraries (not apps!) that can be published to NPM. If you’re looking to build a React-based app, you should use `create-react-app`, `razzle`, `nextjs`, `gatsby`, or `react-static`.

> If you’re new to TypeScript and React, checkout [this handy cheatsheet](https://github.com/sw-yx/react-typescript-cheatsheet/)

## Publishing to Github

Github will trigger a workflow to create a node package everytime a new release is created. The only thing you need to do is create a release and Github will do everything else for you.

## Working with the design system locally

Accessing the latest changes in this repository from external repos before publishing a new package version is possible.

First of all, you need to clone this repo. (Preferrable if it is in the same folder than the external repo)

Now you have to create a symbolic link to your repo:

```bash
yarn link
```

In the external project folder, you have to run this command to create the link between both projects:

```bash
yarn link ../design-system
```

> Please note that the previous command will not work if you cloned the repo in a different directory than the one we recommended in this guide. In that clase, you only need to adjust the `design-system` path and execute the command.

If everything went good, you will find new lines in the `package.json` file indicating the link you just created:

```json
"resolutions": {
    "@zwirner/design-system": "portal:~/SOME-DIRECTORY/design-system"
  }
```

## Commands

TSDX scaffolds your new library inside `/src`, and also sets up a [Parcel-based](https://parceljs.org) playground for it inside `/example`.

The recommended workflow is to run TSDX in one terminal:

```bash
npm start # or yarn start
```

This builds to `/dist` and runs the project in watch mode so any edits you save inside `src` causes a rebuild to `/dist`.

Then run either Storybook or the example playground:

### Storybook

Run inside another terminal:

```bash
yarn storybook
```

This loads the stories from `./stories`.

> NOTE: Stories should reference the components as if using the library, similar to the example playground. This means importing from the root project directory. This has been aliased in the tsconfig and the storybook webpack config as a helper.

### Example

Then run the example inside another:

```bash
cd example
npm i # or yarn to install dependencies
npm start # or yarn start
```

The default example imports and live reloads whatever is in `/dist`, so if you are seeing an out of date component, make sure TSDX is running in watch mode like we recommend above. **No symlinking required**, we use [Parcel's aliasing](https://parceljs.org/module_resolution.html#aliases).

To do a one-off build, use `npm run build` or `yarn build`.

To run tests, use `npm test` or `yarn test`.

## Configuration

Code quality is set up for you with `prettier`, `husky`, and `lint-staged`. Adjust the respective fields in `package.json` accordingly.

### Jest

Jest tests are set up to run with `npm test` or `yarn test`.

### Bundle analysis

Calculates the real cost of your library using [size-limit](https://github.com/ai/size-limit) with `npm run size` and visulize it with `npm run analyze`.

#### Setup Files

This is the folder structure we set up for you:

```txt
/example
  index.html
  index.tsx       # test your component here in a demo app
  package.json
  tsconfig.json
/src
  index.tsx       # EDIT THIS
/test
  blah.test.tsx   # EDIT THIS
/stories
  Thing.stories.tsx # EDIT THIS
/.storybook
  main.js
  preview.js
.gitignore
package.json
README.md         # EDIT THIS
tsconfig.json
```

#### React Testing Library

We do not set up `react-testing-library` for you yet, we welcome contributions and documentation on this.

### Rollup

TSDX uses [Rollup](https://rollupjs.org) as a bundler and generates multiple rollup configs for various module formats and build settings. See [Optimizations](#optimizations) for details.

### TypeScript

`tsconfig.json` is set up to interpret `dom` and `esnext` types, as well as `react` for `jsx`. Adjust according to your needs.

## Continuous Integration

### GitHub Actions

- Every new PR to the `main` branch triggers the `CI` workflow that builds, lints 
and tests the project. Also, every push to the `main` triggers the `CI` workflow.

- Every new PR to the `main` branch triggers the `size` workflow that makes a comment
with cost comparison against the project before the proposed changes. The configuration 
for the `size-limit` can be found in the `package.json`.

## Automatic dependency bumps

The Dependabot configuration is located in the `.github/dependabot.yml`. It is
configured to check weekly for new versions of dependencies in the `package.json`
and actions used in the workflows. If there is a new version, Dependabot opens a PR.

## Optimizations

Please see the main `tsdx` [optimizations docs](https://github.com/palmerhq/tsdx#optimizations). In particular, know that you can take advantage of development-only optimizations:

```js
// ./types/index.d.ts
declare var __DEV__: boolean;

// inside your code...
if (__DEV__) {
  console.log('foo');
}
```

You can also choose to install and use [invariant](https://github.com/palmerhq/tsdx#invariant) and [warning](https://github.com/palmerhq/tsdx#warning) functions.

## Module Formats

CJS, ESModules, and UMD module formats are supported.

The appropriate paths are configured in `package.json` and `dist/index.js` accordingly. Please report if any issues are found.

## Deploying the Example Playground

The Playground is just a simple [Parcel](https://parceljs.org) app, you can deploy it anywhere you would normally deploy that. Here are some guidelines for **manually** deploying with the Netlify CLI (`npm i -g netlify-cli`):

```bash
cd example # if not already in the example folder
npm run build # builds to dist
netlify deploy # deploy the dist folder
```

Alternatively, if you already have a git repo connected, you can set up continuous deployment with Netlify:

```bash
netlify init
# build command: yarn build && cd example && yarn && yarn build
# directory to deploy: example/dist
# pick yes for netlify.toml
```

## Named Exports

Per Palmer Group guidelines, [always use named exports.](https://github.com/palmerhq/typescript#exports) Code split inside your React app instead of your React library.

## Including Styles

There are many ways to ship styles, including with CSS-in-JS. TSDX has no opinion on this, configure how you like.

For vanilla CSS, you can include it at the root directory and add it to the `files` section in your `package.json`, so that it can be imported separately by your users and run through their bundler's loader.

## Usage with Lerna

When creating a new package with TSDX within a project set up with Lerna, you might encounter a `Cannot resolve dependency` error when trying to run the `example` project. To fix that you will need to make changes to the `package.json` file _inside the `example` directory_.

The problem is that due to the nature of how dependencies are installed in Lerna projects, the aliases in the example project's `package.json` might not point to the right place, as those dependencies might have been installed in the root of your Lerna project.

Change the `alias` to point to where those packages are actually installed. This depends on the directory structure of your Lerna project, so the actual path might be different from the diff below.

```diff
   "alias": {
-    "react": "../node_modules/react",
-    "react-dom": "../node_modules/react-dom"
+    "react": "../../../node_modules/react",
+    "react-dom": "../../../node_modules/react-dom"
   },
```

An alternative to fixing this problem would be to remove aliases altogether and define the dependencies referenced as aliases as dev dependencies instead. [However, that might cause other problems.](https://github.com/palmerhq/tsdx/issues/64)


CHANGELOG.md
# v0.0.208 (Fri Dec 15 2023)

#### 🐛 Bug Fix

- animate opacity to show/hide modals [#299](https://github.com/Zwirner/design-system/pull/299) ([@inconduit](https://github.com/inconduit))

#### Authors: 1

- adam ([@inconduit](https://github.com/inconduit))

---

# v0.0.207 (Fri Dec 15 2023)

#### 🐛 Bug Fix

- NWEB-394: Add onClickCTA event to DzPromoModal [#302](https://github.com/Zwirner/design-system/pull/302) ([@grudev888](https://github.com/grudev888))

#### Authors: 1

- [@grudev888](https://github.com/grudev888)

---

# v0.0.206 (Fri Dec 15 2023)

#### 🐛 Bug Fix

- Add unit tests for DzLink component [#307](https://github.com/Zwirner/design-system/pull/307) ([@tabuchid](https://github.com/tabuchid))

#### Authors: 1

- Doug Tabuchi ([@tabuchid](https://github.com/tabuchid))

---

# v0.0.205 (Thu Dec 14 2023)

#### 🐛 Bug Fix

- NWEB-160: Revise Loading State - Simple Version [#298](https://github.com/Zwirner/design-system/pull/298) ([@grudev888](https://github.com/grudev888))

#### Authors: 1

- [@grudev888](https://github.com/grudev888)

---

# v0.0.204 (Thu Dec 14 2023)

#### 🐛 Bug Fix

- Typecheck tests, fix DzFrom.test typechecking, remove broken include [#306](https://github.com/Zwirner/design-system/pull/306) ([@tabuchid](https://github.com/tabuchid))

#### Authors: 1

- Doug Tabuchi ([@tabuchid](https://github.com/tabuchid))

---

# v0.0.203 (Thu Dec 14 2023)

#### 🐛 Bug Fix

- NWEB-1343: add portable text edition to artwork card [#305](https://github.com/Zwirner/design-system/pull/305) ([@JoanCTO](https://github.com/JoanCTO))

#### Authors: 1

- David T ([@JoanCTO](https://github.com/JoanCTO))

---

# v0.0.202 (Thu Dec 14 2023)

#### 🐛 Bug Fix

- NWEB-XYZ: Validate URL's on dzLinks [#304](https://github.com/Zwirner/design-system/pull/304) ([@JoanCTO](https://github.com/JoanCTO))

#### Authors: 1

- David T ([@JoanCTO](https://github.com/JoanCTO))

---

# v0.0.201 (Wed Dec 13 2023)

#### 🐛 Bug Fix

- Nweb 850 fix artist list hover style [#303](https://github.com/Zwirner/design-system/pull/303) ([@inconduit](https://github.com/inconduit))

#### Authors: 1

- adam ([@inconduit](https://github.com/inconduit))

---

# v0.0.200 (Wed Dec 13 2023)

#### 🐛 Bug Fix

- NWEB-875: add regex validation for internal links [#291](https://github.com/Zwirner/design-system/pull/291) ([@JoanCTO](https://github.com/JoanCTO) [@tabuchid](https://github.com/tabuchid))

#### Authors: 2

- David T ([@JoanCTO](https://github.com/JoanCTO))
- Doug Tabuchi ([@tabuchid](https://github.com/tabuchid))

---

# v0.0.199 (Wed Dec 13 2023)

#### 🐛 Bug Fix

- NWEB-1447: capitalize text on form input [#301](https://github.com/Zwirner/design-system/pull/301) ([@JoanCTO](https://github.com/JoanCTO))

#### Authors: 1

- David T ([@JoanCTO](https://github.com/JoanCTO))

---

# v0.0.198 (Wed Dec 13 2023)

#### 🐛 Bug Fix

- First form component test [#294](https://github.com/Zwirner/design-system/pull/294) ([@inconduit](https://github.com/inconduit))

#### Authors: 1

- adam ([@inconduit](https://github.com/inconduit))

---

# v0.0.197 (Wed Dec 13 2023)

#### 🐛 Bug Fix

- NWEB-931: Regression: Navigation Behavior is broken [#284](https://github.com/Zwirner/design-system/pull/284) ([@grudev888](https://github.com/grudev888))

#### Authors: 1

- [@grudev888](https://github.com/grudev888)

---

# v0.0.196 (Wed Dec 13 2023)

#### 🐛 Bug Fix

- NWEB-1291: Popup image fixes [#290](https://github.com/Zwirner/design-system/pull/290) ([@MadaShindeInai](https://github.com/MadaShindeInai))
- chore: set NextImage as unoptimised [#297](https://github.com/Zwirner/design-system/pull/297) ([@MadaShindeInai](https://github.com/MadaShindeInai))

#### Authors: 1

- Sergei ([@MadaShindeInai](https://github.com/MadaShindeInai))

---

# v0.0.195 (Tue Dec 12 2023)

#### 🐛 Bug Fix

- NWEB-1373: update Newsletter Pop Up text [#296](https://github.com/Zwirner/design-system/pull/296) ([@JoanCTO](https://github.com/JoanCTO))
- NWEB-1387: fix 6col title size for content cards [#295](https://github.com/Zwirner/design-system/pull/295) ([@JoanCTO](https://github.com/JoanCTO))
- NWEB-1350: [QA Feedback] Regression: Inquiry and Newsletter form "Terms of Service" Direction [#289](https://github.com/Zwirner/design-system/pull/289) ([@grudev888](https://github.com/grudev888))

#### Authors: 2

- [@grudev888](https://github.com/grudev888)
- David T ([@JoanCTO](https://github.com/JoanCTO))

---

# v0.0.194 (Mon Dec 11 2023)

#### 🐛 Bug Fix

- NWEB-650: Hide alphabetical list on artist list page upon initial page load [#286](https://github.com/Zwirner/design-system/pull/286) ([@grudev888](https://github.com/grudev888))

#### Authors: 1

- [@grudev888](https://github.com/grudev888)

---

# v0.0.193 (Fri Dec 08 2023)

#### 🐛 Bug Fix

- NWEB-XYZ: Update newsletter form buttons and terms markup according figma [#287](https://github.com/Zwirner/design-system/pull/287) ([@MadaShindeInai](https://github.com/MadaShindeInai))

#### Authors: 1

- Sergei ([@MadaShindeInai](https://github.com/MadaShindeInai))

---

# v0.0.192 (Fri Dec 08 2023)

#### 🐛 Bug Fix

- upgrade deps to match web dep versions [#285](https://github.com/Zwirner/design-system/pull/285) ([@inconduit](https://github.com/inconduit))

#### Authors: 1

- adam ([@inconduit](https://github.com/inconduit))

---

# v0.0.191 (Thu Dec 07 2023)

#### 🐛 Bug Fix

- NWEB-1317-2: add overlay to interstitials [#280](https://github.com/Zwirner/design-system/pull/280) ([@JoanCTO](https://github.com/JoanCTO))

#### Authors: 1

- David T ([@JoanCTO](https://github.com/JoanCTO))

---

# v0.0.190 (Thu Dec 07 2023)

#### ⚠️ Pushed to `main`

- Update .autorc ([@tabuchid](https://github.com/tabuchid))
- Update chromatic.yml ([@tabuchid](https://github.com/tabuchid))

#### 🔩 Dependency Updates

- Update chromatic.yml [#275](https://github.com/Zwirner/design-system/pull/275) ([@tabuchid](https://github.com/tabuchid))

constants/mocks/DzTitle.ts
export const pageData = {
  category: 'Category Title',
  title: 'Lorem ipsum dolor sit amet, consectetuer adipiscin',
  subtitle: 'Lorem ipsum dolor sit amet, consectetuer adipiscin',
  description:
    'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa quis enim. Donec pede justo, fringilla vel, aliquet nec, vulputate eget, arcu. In enim justo, rhoncus ut, imperdiet a, venenatis vitae, justo. Nullam dictum felis eu pede mollis pretium. Integer tincidunt. Cras dapibu',
  primaryCTA: {
    title: 'Button',
    description: 'Descriptive Title',
  },
};

export const sectionData = {
  title: 'Lorem ipsum dolor sit amet, consectetuer adipiscin',
  subtitle: 'Lorem ipsum dolor sit amet, consectetuer adipiscin',
  linkCTA: {
    text: 'View All',
    url: '/',
  },
};

export const exhibitionData = {
  title: 'Sherrie Levine: Wood',
  openingReception: 'April 19, 6-9pm',
  startDate: '2023-08-06',
  endDate: '2023-08-29',
  pressReleasePDFURL:
    'https://cdn.sanity.io/files/juzvn5an/test/3117e550d45fd4effbcd064d048cf9768f2298d0.pdf',
  checklistPDFURL:
    'https://cdn.sanity.io/files/juzvn5an/test/3117e550d45fd4effbcd064d048cf9768f2298d0.pdf',
  artists: [
    {
      description:
        "Katherine Bernhardt (b. 1975) was born in St. Louis, Missouri and received a BFA from the School of the Art Institute of Chicago in 1998 and an MFA from the School of Visual Arts, New York, in 2000.  \n\nIn 2018, the solo exhibition Katherine Bernhardt: Watermelon World was on view at the Mario Testino Museum (MATE) in Lima, Peru. The previous year, in 2017, the Modern Art Museum of Fort Worth, Texas presented FOCUS: Katherine Bernhardt. Also in 2017, the artist painted a 60 foot-long mural entitled XXL Superflat Pancake for the St. Louis Contemporary Art Museum. Bernhardt has created a permanent installation for Club Caribe, Cidra, Puerto Rico and a pool painting at the Nautilus South Beach, Miami Beach (both 2015).\n \nHer work has also been included in significant group exhibitions, such as We Fight to Build a Free World: An Exhibition by Jonathan Horowitz, Jewish Museum, New York (2020); Animal Farm, an exhibition curated by Sadie Laska at the Brant Foundation Art Study Center, Greenwich, Connecticut (2017); NO MAN’S LAND: Women Artists from the Rubell Family Collection, Rubell Museum, Miami, which traveled to the National Museum of Women in the Arts, Washington, DC (2015-2017); and Bad Touch, Ukrainian Institute of Modern Art, Chicago (2002). \n\nOpening May 20, 2023 at the gallery's Hong Kong location is a presentation of new works by the artist. In 2022, David Zwirner, London presented an exhibition of new work by the artist, Katherine Bernhardt: why is a mushroom growing in my shower?\n\nWork by the artist is found in prominent public and museum collections, including the Brant Foundation Art Study Center, Greenwich, Connecticut; Carnegie Museum of Art, Pittsburgh; Fondazione Sandretto Re Rebaudengo, Turin, Italy; High Museum of Art, Atlanta; Hirshhorn Museum and Sculpture Garden, Washington, DC; Knoxville Museum of Art, Tennessee; Portland Museum of Art, Maine; Rubell Museum, Miami; and San Antonio Museum of Art, Texas. Bernhardt lives and works in St. Louis. \n\n\nClick here to download full CV\n",
      _updatedAt: '2023-07-21T23:13:38Z',
      _type: 'artist',
      _rev: 'xqt72NIuqWVl36Qb8SPady',
      _createdAt: '2023-05-12T18:55:46Z',
      _id: '018151B4-5A34-4011-9168-0F6CF3BCBC92',
      birthdate: '',
      fullName: 'Katherine Bernhardt',
      picture: {
        _type: 'image',
        asset: {
          _ref: 'image-8c6991ae442f7644be682b03cbddf13bef25d599-2056x1196-jpg',
          _type: 'reference',
        },
      },
      affiliation: true,
      lastName: 'Bernhardt',
      url: 'https://www.davidzwirner.com/artists/katherine-bernhardt',
      firstName: 'Katherine',
      summary:
        'Katherine Bernhardt (b. 1975) was born in St. Louis, Missouri and received a BFA from the School of the Art Institute …',
    },
    {
      fullName: 'Josh Smith',
      artistPage: {
        slug: {
          current: '/artists/josh-smith',
          _type: 'slugUrl',
        },
      },
    },
  ],
  location: {
    name: 'New York: 52 Walker',
    _id: '63b5ee48-633f-4ef8-87d6-6f7174bea072',
    type: 'dz-gallery',
    _updatedAt: '2023-07-21T23:13:39Z',
    url: 'https://www.52walker.com/',
    timezone: 'America/New_York',
    _createdAt: '2023-05-12T19:49:56Z',
    _rev: 'xqt72NIuqWVl36Qb8SPtBi',
    _type: 'location',
    photos: [
      {
        _type: 'image',
        _key: 'a3d7c298d494',
        asset: {
          _ref: 'image-847168274a93a2d8551d3f85310d90ad70b19fb2-2459x1720-jpg',
          _type: 'reference',
        },
      },
    ],
    hours: [
      {
        day: 'Monday',
        availableTimes: [],
        _type: 'availabilityDay',
        _key: 'monday',
      },
      {
        _key: 'tuesday',
        day: 'Tuesday',
        availableTimes: [
          {
            from: '10:00 AM',
            to: '6:00 PM',
            _key: '4494f502f855',
            _type: 'availabilityDuration',
          },
        ],
        _type: 'availabilityDay',
      },
      {
        _key: 'wednesday',
        day: 'Wednesday',
        availableTimes: [
          {
            to: '6:00 PM',
            _key: '4494f502f855',
            _type: 'availabilityDuration',
            from: '10:00 AM',
          },
        ],
        _type: 'availabilityDay',
      },
      {
        _type: 'availabilityDay',
        _key: 'thursday',
        day: 'Thursday',
        availableTimes: [
          {
            _type: 'availabilityDuration',
            from: '10:00 AM',
            to: '6:00 PM',
            _key: '4494f502f855',
          },
        ],
      },
      {
        _type: 'availabilityDay',
        _key: 'friday',
        day: 'Friday',
        availableTimes: [
          {
            _type: 'availabilityDuration',
            from: '10:00 AM',
            to: '6:00 PM',
            _key: '4494f502f855',
          },
        ],
      },
      {
        _key: 'saturday',
        day: 'Saturday',
        availableTimes: [
          {
            _type: 'availabilityDuration',
            from: '10:00 AM',
            to: '6:00 PM',
            _key: '887f6c231a4e',
          },
        ],
        _type: 'availabilityDay',
      },
      {
        _type: 'availabilityDay',
        _key: 'sunday',
        day: 'Sunday',
        availableTimes: [
          {
            _key: '887f6c231a4e',
            _type: 'availabilityDuration',
            from: '10:00 AM',
            to: '6:00 PM',
          },
        ],
      },
    ],
    address: {
      _type: 'address',
      addressLine2: 'Address Line 2',
      state: 'New York',
      addressLine: '52 Walker Street',
      country: 'United States',
      zipCode: '10013',
      city: 'New York',
    },
  },
};
